File: src/problemset/array/RotateArray.java
package problemset.array;

/*
 * @date_time: 2020-04-03 23:32
 * @author: MoeexT
 * @problem: 189
 * @url: https://leetcode-cn.com/problems/rotate-array/
 */

import java.util.Arrays;

public class RotateArray {
    /*
     * @time: 50 ms 35.28%
     * @mem: 40.8MB 5.09%
     */
    public void rotateStepByStep(int[] nums, int k) {
        int len = nums.length;
        if (len == 0) {
            return;
        }
        for (int i = 0; i < k; i++) {
            int tmp = nums[len - 1];
            System.arraycopy(nums, 0, nums, 1, len - 1);
            nums[0] = tmp;
        }

    }

    public void rotate(int[] nums, int k) {
        int len = nums.length;
        if (len == 0) {
            return;
        }

        int src = len - 1, tgt = k - 1;
        int tmp1, tmp2 = nums[src];

        while (src >= k) {
            tmp1 = nums[(len + tgt) % len];
            nums[(len + tgt) % len] = tmp2;
            tgt--;
            src--;
            tmp2 = nums[src];
            nums[src] = tmp1;
//            System.out.println(tmp1+", "+tmp2);
        }

    }

    public static void main(String[] args) {
        RotateArray solution = new RotateArray();
        int[] nums = {1, 2, 3, 4, 5, 6, 7, 8};
        solution.rotate(nums, 3);
        System.out.println(Arrays.toString(nums));
    }
}

File: src/problemset/array/RotateArray.go
package main

import "fmt"

// url: https://leetcode-cn.com/problems/rotate-array/solution/
// problem: 189

func rotate(nums []int, k int) {
	n := len(nums)
	k = k % n
	if k == 0 {
		return
	}

	if n&1 == 0 && k&1 == 0 {
		vessle, target := nums[0], k
		for i := 0; i < n/2; i++ {
			nums[target], vessle = vessle, nums[target]
			target = (target + k) % n
		}
		vessle, target = nums[1], k+1
		for i := 0; i < n/2; i++ {
			nums[target], vessle = vessle, nums[target]
			target = (target + k) % n
		}
	} else {
		vessle, target := nums[0], k
		for i := 0; i < n; i++ {
			nums[target], vessle = vessle, nums[target]
			target = (target + k) % n
		}
	}
	fmt.Println(nums)
}

func main() {
	rotate([]int{1, 2, 3, 4, 5, 6}, 3)
	/**
	[1,2,3,4,5,6]
	[3,4,5,6,1,2]
	[1,2,3,4,5,6,7]
	[6,7,1,2,3,4,5]
	*/
}
